### PR TITLE
AAS Windows Universal Extension

### PR DESCRIPTION
**Pull request by avedmala**
This PR adds an Azure Site Extension for supporting multiple runtimes Windows Azure App Service environment. The goal is to support Node, .NET, Java and Python.

This extension will be universal because it's able to download and instrument the appropriate tracer by determining the runtime during extension install/startup.

There will also be an IIS module bundled in the extension which will manage the agents' processes from within the same sandbox as the main app. This is the main problem we're solving as we can't kick off these processes with WebJobs and modifying each tracer to launch the agents will be a lot of upstream changes across multiple repos.

### CODE REVIEW

File: windows/AgentProcessManager/AgentProcessManager.cpp [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+
+    int StartAgent(const std::wstring &agentName)
+    {
+        /*
+        This prevents a race condition where two threads from a .NET app are
+        triggered from two user requests at the same time. Being the first
+        request, they'll both bypass the previous check. The lock will ensure
+        that only the first request spawns an agent.
+        */
+        std::lock_guard<std::mutex> lock(mtx);
```
Comment (avedmala) [inline windows/AgentProcessManager/AgentProcessManager.cpp:78]: Definitely need a second look on if I'm using this lock properly...

File: windows/AgentProcessManager/AgentProcessManager.cpp [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+        StartAgent(L"dogstatsd");
+        return GL_NOTIFICATION_CONTINUE;
+    }
+
+    GLOBAL_NOTIFICATION_STATUS OnGlobalPreBeginRequest(IN IPreBeginRequestProvider *pProvider)
+    {
+        UNREFERENCED_PARAMETER(pProvider);
+        if (!ProcessExists("trace-agent"))
+        {
+            HANDLE hMutexA = CreateMutex(NULL, TRUE, L"trace-agent");
```
Comment (avedmala) [inline windows/AgentProcessManager/AgentProcessManager.cpp:44]: Need a close look at this locking logic to make sure I'm doing it right

File: windows/AgentProcessManager/AgentProcessManager.cpp [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+        StartAgents();
+        return GL_NOTIFICATION_CONTINUE;
+    }
+
+    GLOBAL_NOTIFICATION_STATUS OnGlobalPreBeginRequest(IN IPreBeginRequestProvider *pProvider)
+    {
+        UNREFERENCED_PARAMETER(pProvider);
+        if (!ProcessExists("process_manager"))
+        {
+            HANDLE hMutex = CreateMutex(NULL, TRUE, L"process_manager");
```
Comment (avedmala) [inline windows/AgentProcessManager/AgentProcessManager.cpp:42]: Def need a close look on this logic to make sure it's correct

**Comment by avedmala** [comment]
The commit history is a bit messy but the current state is the build script can make Node and Java extensions (there's code for Dotnet but we need the *.csproj files). If we use `--debug`, it'll make it for the just Windows extension.

Latest Windows Extension is actually just a Java one because I wanted to test it but didn't want to overwrite the existing Java extension.

**Comment by duncanpharvey** [comment]
Closing in favor of PR [Node Extension GA](https://github.com/DataDog/datadog-aas-extension/pull/229)

**Pull request closed without merging**